lxd/profiles_utils.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/lxc/lxd/shared/api"
+)
+
+func doProfileUpdate(d *Daemon, name string, id int64, profile *api.Profile, req api.ProfilePut) Response {
+	// Sanity checks
+	err := containerValidConfig(d, req.Config, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	err = containerValidDevices(req.Devices, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	containers := getContainersWithProfile(d, name)
+
+	// Check if the root device is supposed to be changed or removed.
+	oldProfileRootDiskDeviceKey, oldProfileRootDiskDevice := containerGetRootDiskDevice(profile.Devices)
+	_, newProfileRootDiskDevice := containerGetRootDiskDevice(req.Devices)
+	if len(containers) > 0 &&
+		oldProfileRootDiskDevice["pool"] != "" &&
+		newProfileRootDiskDevice["pool"] == "" ||
+		(oldProfileRootDiskDevice["pool"] != newProfileRootDiskDevice["pool"]) {
+		// Check for containers using the device
+		for _, container := range containers {
+			// Check if the device is locally overridden
+			localDevices := container.LocalDevices()
+			k, v := containerGetRootDiskDevice(localDevices)
+			if k != "" && v["pool"] != "" {
```

Comment:
So I think this is what we want here:
- containerValidDevices() checks that there are no entries with the same path.
- now we check for a local device in the containers local devices map.
- if we find local device with an appropriate pool, we go to the next container
- otherwise we do the check you suggested, looking whether there is another profile that provides the same root disk device

I just want to add another test case to the suite, that we cannot assign profiles to a container that specify conflicting root disk devices when it relies on a profile for its root disk device.